ratis-server/src/main/java/org/apache/ratis/server/impl/DataStreamServerImpl.java

```diff
@@ -35,31 +37,38 @@
 public class DataStreamServerImpl implements DataStreamServer {
   public static final Logger LOG = LoggerFactory.getLogger(DataStreamServerImpl.class);
 
-  private final DataStreamServerRpc serverRpc;
+  private Optional<DataStreamServerRpc> serverRpc;
```

Comment:
We do not need to change DataStreamServerImpl at all.  The idea is to implement a new RPC type for DISABLED, i.e. we should implement a DisabledDataStreamFactory.
```
public class DisabledDataStreamFactory implements DataStreamServerFactory, DataStreamClientFactory {
  public DisabledDataStreamFactory(Parameters parameters){}

  @Override
  public SupportedDataStreamType getDataStreamType() {
    return SupportedDataStreamType.DISABLED;
  }

  @Override
  public DataStreamClientRpc newDataStreamClientRpc(RaftPeer server, RaftProperties properties) {
    return new DataStreamClientRpc() {
      @Override
      public void startClient() {}

      @Override
      public void closeClient() {}
    };
  }

  @Override
  public DataStreamServerRpc newDataStreamServerRpc(RaftPeer server, StateMachine stateMachine,
      RaftProperties properties) {
    return new DataStreamServerRpc() {
      @Override
      public void start() {}

      @Override
      public void close() {}

      @Override
      public void addRaftPeers(Collection<RaftPeer> peers) {}
    };
  }

  @Override
  public DataStreamServerRpc newDataStreamServerRpc(RaftServer server, StateMachine stateMachine,
      RaftProperties properties) {
    return new DataStreamServerRpc() {
      @Override
      public void start() {}

      @Override
      public void close() {}

      @Override
      public void addRaftPeers(Collection<RaftPeer> peers) {}
    };
  }
}
```

Reply:
Thanks! 

:) somehow I thought this approach complicates the code too much but I was wrong. This approach encapsulates code much well thus simplify the code.